Development auth part

diaabadr opened the pull request:
## Describtion:
Working on Authentication and email sender
## Issues
#7 #15

diaabadr committed:
user model creating and updating

diaabadr committed:
adding signup handler

diaabadr committed:
modifying User

diaabadr committed:
adding jwt

diaabadr committed:
.gitignore
middleware
handlers
tests

diaabadr committed:
Delete .env

diaabadr committed:
gitignore

diaabadr committed:
added tests for: update, get, refreshToken, signin

diaabadr committed:
adding workflows

diaabadr committed:
adding development to prs workflow triggering

diaabadr committed:
workflow file update

diaabadr committed:
adding missed check on error while migration

diaabadr committed:
updating modules

diaabadr committed:
updating workflow

diaabadr committed:
documenting

diaabadr committed:
formatting user.go

rawdaGastan commented on .github/workflows/flyspray.yaml:
```diff
@@ -0,0 +1,41 @@
+name: Flyspray Test
```
you weren't supposed to add a workflow. It has another issue

rawdaGastan commented on backend/app/app.go:
```diff
@@ -30,9 +31,21 @@ type App struct {
-	a.router = gin.Default()
 
 	// set routes here
+	a.setUserRoutes()
```
it should be setRoutes. as it will include all routes

rawdaGastan commented on backend/app/app.go:
```diff
@@ -30,9 +31,21 @@ type App struct {
+	userGroup.POST("/signin", WrapFunc(a.SignIn))
+	userGroup.POST("/signup/verify", WrapFunc(a.Verify))
+	userGroup.POST("/refresh_token", WrapFunc(a.RefreshToken))
+	userGroup.PUT("", middleware.RequireAuth, WrapFunc(a.UpdateUser))
```
you can make your router use a specific middlewares instead of adding it with each route. so any one can use your auth_route and add his routes

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	"github.com/gin-gonic/gin"
+)
+
+func (a *App) Signup(ctx *gin.Context) (interface{}, Response) {
```
no need for handlers to be exported
apply for all

rawdaGastan replied:
you should handle the case if user is created but no verified yet.

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+)
+
+func (a *App) Signup(ctx *gin.Context) (interface{}, Response) {
+	var user models.User
```
why you don't add input structs for your api?

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+
+	err := json.NewDecoder(ctx.Request.Body).Decode(&user)
+
+	if err != nil {
```
please log your errors. zerolog is recommended
apply for all

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	err := json.NewDecoder(ctx.Request.Body).Decode(&user)
+
+	if err != nil {
+		return nil, BadRequest(err)
```
you should return an error that the user will understand. `input data is invalid` for example

apply for all

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, BadRequest(err)
+	}
+
+	user.Password = hash
```
you should take password and confirm password from the user

diaabadr replied:
i think that check would be better in the client side. what do you see?

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	user, err = a.client.CreateUser(user)
+
+	if err != nil {
+		if err.Error() == "UNIQUE constraint failed: users.email" {
```
I think you can check if user exists or not first instead of comparing errors

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, BadRequest(err)
+	}
+
+	result := a.client.Client.Model(&models.User{}).Where("Verification_code = ?", requestBody.VerificationCode).Update("Verified", true)
```
db logic should be moved to db pkg

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	return ResponseMsg{Message: message}, Created()
+}
+
+func (a *App) Verify(ctx *gin.Context) (interface{}, Response) {
```
you should:
- check if user is created.
- if user is already verified
- use update time to save your code update time. if timeout then the code has expired

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+
+func (a *App) SignIn(ctx *gin.Context) (interface{}, Response) {
+
+	var requestBody models.User
```
sign in input isn't a user model. It's just an email and a password

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	var user models.User
+	result := a.client.Client.First(&user, "Email = ?", requestBody.Email)
```
move DB logic to functions in the db pkg

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		if result.Error == gorm.ErrRecordNotFound {
+			return nil, NotFound(errors.New("invalid email or password"))
+		}
+		return nil, InternalServerError(result.Error)
```
bad request

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+
+	if !isPasswordMatches {
+
+		return nil, NotFound(errors.New("invalid email or password"))
```
bad request

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, InternalServerError(err)
+	}
+
+	refreshToken, err := internal.GenerateRefreshToken(user)
```
why generating a refresh token for sign in?

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	if !user.Verified {
+		return nil, Forbidden(errors.New("account not verified"))
```
what about `account is not verified yet, please check the verification email in your inbox`?

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		Verified: user.Verified,
+	}
+
+	ctx.SetSameSite(http.SameSiteLaxMode)
```
no need for it

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	ctx.SetSameSite(http.SameSiteLaxMode)
+	ctx.SetCookie("Authorization", accessToken, 60*15, "", "", true, true)
```
we don't use cookies

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	return ResponseMsg{Message: "logged in successfully", Data: struct {
+		AccessToken  string      `json:"access_token"`
+		RefreshToken string      `json:"refresh_token"`
+		User         models.User `json:"user"`
```
no need to return `user`  and  `refreshToken`

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	userPayload, exists := ctx.Get("user")
+
+	if !exists {
+		return nil, UnAuthorized(errors.New("user not found"))
```
`user is not found`
please check all

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, UnAuthorized(errors.New("user not found"))
+	}
+
+	var user models.User
```
would you accept an ID update? we shouldn't take a `user model` from the user

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, InternalServerError(err)
+	}
+
+	_, err := a.client.GetUserByID(user.ID)
```
user ID should be from your payload

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, InternalServerError(err)
+	}
+
+	return ResponseMsg{Message: "updated successfully"}, Created()
```
return ok() it's an update

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	return ResponseMsg{Message: "found", Data: user}, Ok()
+}
+
+func (a *App) RefreshToken(ctx *gin.Context) (interface{}, Response) {
```
you should take the access token from your Authorization in headers and it will be a bearer token

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+
+func (a *App) RefreshToken(ctx *gin.Context) (interface{}, Response) {
+
+	var body struct {
```
we shouldn't take an input from the user in this api

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+		return nil, InternalServerError(fmt.Errorf("error refreshing token: %v", err))
+	}
+
+	ctx.SetSameSite(http.SameSiteLaxMode)
```
no need

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,246 @@
+	return ResponseMsg{
+		Message: "token has been refreshed successfully",
+		Data: struct {
+			AccessToken string `json:"access_token"`
```
you should return the refreshed token

rawdaGastan commented on backend/app/wrapper.go:
```diff
@@ -0,0 +1,154 @@
+
+// InternalServerError result
+func InternalServerError(err error) Response {
+	return Error(err, 500)
```
this should be 0 not 500 beacuse it is the default in the error function

rawdaGastan commented on backend/app/wrapper.go:
```diff
@@ -0,0 +1,154 @@
+
+// Conflict response
+func Conflict(err error) Response {
+	return Error(err, 409)
```
please use the status name instead of 409

rawdaGastan commented on backend/cmd/server.go:
```diff
@@ -5,9 +5,15 @@ import (
 
 func main() {
+
+	err := internal.LoadEnv()
```
you are not supposed to make the configs.

rawdaGastan commented on backend/internal/config.go:
```diff
@@ -0,0 +1,8 @@
+package internal
```
config is in another issue please remove it

rawdaGastan commented on backend/internal/jwt.go:
```diff
@@ -0,0 +1,78 @@
+	Email    string
+	Name     string
+	ID       string
+	Verified bool
```
no need for verified and name

rawdaGastan commented on backend/internal/jwt.go:
```diff
@@ -0,0 +1,78 @@
+// GenerateAccessToken generates an access token
+func GenerateAccessToken(user models.User) (string, error) {
+
+	expirationDate := time.Now().Add(time.Minute * 15)
```
timeout should't be hardcoded

rawdaGastan commented on backend/internal/jwt.go:
```diff
@@ -0,0 +1,78 @@
+
+// GenerateRefreshToken generates a refresh token
+func GenerateRefreshToken(user models.User) (string, error) {
+	expirationDate := time.Now().Add(time.Hour * 24 * 3)
```
hardcoding?

rawdaGastan commented on backend/internal/jwt.go:
```diff
@@ -0,0 +1,78 @@
+		return []byte(os.Getenv("SECRET")), nil
+	})
+
+	if err != nil {
```
you can check if `token.valid`

rawdaGastan commented on backend/internal/jwt.go:
```diff
@@ -0,0 +1,78 @@
+		return nil, errors.New("invalid token")
+	}
+
+	if time.Now().After((*claims.ExpiresAt).Time) {
```
you are checking if now is after the expiration time? it will be always expired

rawdaGastan commented on backend/internal/password_hashing.go:
```diff
@@ -0,0 +1,16 @@
+package internal
+
+import "golang.org/x/crypto/bcrypt"
```
- first we should use salt
the SALT shouldn't be passed to the hash function NOR it should be chosen or set in the config at all
the salt should be totally randomly generated before computing the hash. hence it will be unique per password
The idea behind the salt to make it impossible to do a hashtable attack here is how it works and how it should be used:
- If we have passowrd x, and hash(x) = y
- it is impossible to compute x back from y, in other words fn(y) = x does not exist.
- now, someone can create a huge table of all common passwords like `password`, `123456`, etc.. then run a hash function
  on this huge list, and create a hash table where `k = hash(x), v=x`.
- this mean that if I now have the hash `y`, and IF y is one of the easy common password, i can simply do look up `y` in the
  hash table, and then get `x`` since all hashes where pre computed by someone.
Now, to avoid this we use salt: and this is how salt work (and why this function and the way you configure salt is totally wrong)
- I generate salt=random() of fixed length say 10 bytes (every time)
- I then compute y = hash(salt + x)
- Then i create the hashed password (the one i actually store) as `hashed = salt + y` (this can be then hex encoded or base64 it does not matter)
- Now it's IMPOSSIBLE for an attacker to use a hash table attach on the stored `hashed` because even if password is `123456` it will never be found
  on the hash table, right ?
- Now you are wondering how then i can validate the password against hashed, it's very simple
- I can have validate function that looks like this validate(hashed, password)
- Then, it can do this `(salt, y) = (hashed[:10], hashed[10:])`
- Then simply do `return y == hash(salt + x)`
Also notice that also means if more than one user used the same password, their stored hashes will look completely different it means even
if one password is cracked, the other users using the same password won't be affected.

diaabadr replied:
Really good to know. but i found that bcrypt uses salts by default in its internal implementation

rawdaGastan commented on backend/middlewares/middlewares.go:
```diff
@@ -0,0 +1,45 @@
+		ctx.Abort()
+		return
+	}
+	ctx.Set("user", string(userEncoded))
```
only set user ID

rawdaGastan commented on backend/middlewares/middlewares.go:
```diff
@@ -0,0 +1,45 @@
+
+// RequireAuth is a middleware that checks if the user is authenticated
+func RequireAuth(ctx *gin.Context) {
+	tokenString, err := ctx.Cookie("Authorization")
```
don't use cookies, use authorization on headers with bearer token option

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+	Email            string `json:"email" gorm:"unique;not null" validate:"regexp=^[0-9a-z]+@[0-9a-z]+(\\.[0-9a-z]+)+$"`
+	Password         string `json:"password" gorm:"not null"`
+	VerificationCode string `gorm:"unique;not null"`
+	Verified         bool   `gorm:"default:false"`
```
add update time

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// CreateUser creates a user
+func (db *DBClient) CreateUser(user User) (User, error) {
```
check `BeforeCreate` function to automate the creation of the IDs

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+func (db *DBClient) GetUserByID(id string) (User, error) {
+	var user User
+
+	result := db.Client.Select("id", "email", "name", "verified").First(&user)
```
no need for select. add your condition in the First function

diaabadr replied:
to avoid returning password

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+
+// UpdateUser updates a user
+func (db *DBClient) UpdateUser(user User) error {
+	if user.Password != "" {
```
checks and validations must be outside db logic

diaabadr replied:
here i'm checking if he wants to reset the password. that's not validation

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+// GenerateVerificationCode generates a unique verification code
+func (db *DBClient) GenerateVerificationCode() string {
+
+	charsSet := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
```
should be numbers only. check `rand.Seed`

rawdaGastan commented on backend/models/user.go:
```diff
@@ -0,0 +1,91 @@
+	Name             string `json:"name"`
+	Email            string `json:"email" gorm:"unique;not null" validate:"regexp=^[0-9a-z]+@[0-9a-z]+(\\.[0-9a-z]+)+$"`
+	Password         string `json:"password" gorm:"not null"`
+	VerificationCode string `gorm:"unique;not null"`
```
code should be int?

rawdaGastan commented:
Please add your issue to the pr description. move the status of the issue to `in progress`

diaabadr committed:
modyfying signin handler to just return access token
modyfying refresh_token endpoint
adding check for unverified accounts
logging errors
adding input structs
changing some error messages
moving dblogic to DB
adding timeout for verification code

diaabadr committed:
modifying tests
adding tests for new cases
adding any .db to gitignore

diaabadr committed:
adding uniqueness check on update for more readable error

diaabadr committed:
Merge branch 'development' into development_auth_part

diaabadr committed:
solving conflicts

diaabadr committed:
formatting

rawdaGastan commented on backend/app/app.go:
```diff
@@ -59,4 +58,20 @@ func (app *App) setRoutes() {
 		comment.PUT("/:id", WrapFunc(app.updateComment))
 	}
 
+	authUserGroup := app.router.Group("/user")
```
you should't link your auth to the user only. you should have an auth group that can be used to create other groups using auth. please modify comment and project parts to use auth

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+	err := json.NewDecoder(ctx.Request.Body).Decode(&requestBody)
+
+	if err != nil {
+		log.Error().Err(err).Msg("")
```
you can use `Send()` instead of `Msg("")`

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+)
+
+type signupBody struct {
+	Name            string `json:"name"`
```
mark all as required
check sign in input too

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+	Email           string `json:"email"`
+	Password        string `json:"password"`
+	ConfirmPassword string `json:"confirm_password"`
+	Verified        bool   `json:"verified"`
```
why taking verified from the user?

diaabadr committed:
Delete .github/workflows directory

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+	err = user.Validate()
+	if err != nil {
+		log.Error().Err(err).Msg("")
+		return nil, BadRequest(fmt.Errorf("validation error: %w", err))
```
you should't return err msgs to the user. `invalid user data` is good

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+		Email:                   requestBody.Email,
+		Password:                requestBody.Password,
+		Verified:                requestBody.Verified,
+		VerificationCodeTimeout: time.Now().Add(time.Hour * 2),
```
don't hard code it. 2 hours are too long also.

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+
+	if err != nil {
+		log.Error().Err(err).Msg("")
+		return nil, BadRequest(fmt.Errorf("failed to hash password: %w", err))
```
internal server error

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,287 @@
+	user.Password = hash
+
+	user.VerificationCode = a.DB.GenerateVerificationCode()
+	fmt.Println(user.VerificationCode)
```
remove print

diaabadr committed:
Delete .github/workflows directory

diaabadr committed:
Merge branch 'development_auth_part' of https://github.com/codescalersinternships/Flyspray into development_auth_part

diaabadr committed:
adding workflow

diaabadr committed:
refactoring

diaabadr committed:
adding new test
refactoring old tests

diaabadr committed:
adding test to verify

diaabadr committed:
Merge branch 'development' into development_auth_part

diaabadr committed:
triggering workflow

diaabadr committed:
adding tests to verify

diaabadr committed:
formatting

diaabadr committed:
adding member routes to auth middleware

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,336 @@
+			err = a.DB.UpdateVerificationCode(user.ID, verifivationCode, timeout)
+			if err != nil {
+				log.Error().Err(err).Send()
+				return "", InternalServerError(errInternalServerError)
```
return nil not "", check all

rawdaGastan commented on backend/app/auth_handlers.go:
```diff
@@ -0,0 +1,336 @@
+	}
+
+	fmt.Println(user.VerificationCodeTimeout)
+	fmt.Println(time.Now().Add(time.Duration(0)))
```
remove prints

rawdaGastan commented on backend/cmd/server.go:
```diff
@@ -5,9 +5,16 @@ import (
 
 func main() {
+
+	err := godotenv.Load()
```
where is it used?

diaabadr committed:
removing prints
removing godotenv

diaabadr committed:
formatting
